src/python/pants/backend/python/rules/repl.py

```diff
@@ -1,69 +1,86 @@
 # Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 
-from pants.backend.python.rules.pex import Pex
-from pants.backend.python.rules.pex_from_targets import PexFromTargetsRequest
+from pants.backend.python.rules.pex import Pex, PexRequest, PexRequirements
+from pants.backend.python.rules.pex_from_targets import (
+    PexFromTargetsRequest,
+    requirements_pex_from_targets_request,
+)
 from pants.backend.python.rules.python_sources import PythonSourceFiles, PythonSourceFilesRequest
 from pants.backend.python.subsystems.ipython import IPython
-from pants.backend.python.target_types import PythonSources
 from pants.core.goals.repl import ReplImplementation, ReplRequest
+from pants.engine.addresses import Addresses
 from pants.engine.fs import Digest, MergeDigests
 from pants.engine.rules import Get, MultiGet, collect_rules, rule
 from pants.engine.unions import UnionRule
 
 
 class PythonRepl(ReplImplementation):
     name = "python"
-    required_fields = (PythonSources,)
 
 
 @rule
 async def create_python_repl_request(repl: PythonRepl) -> ReplRequest:
-    pex_request = Get(
+    requirements_request = Get(
         Pex,
-        PexFromTargetsRequest(
-            (tgt.address for tgt in repl.targets),
-            output_filename="python.pex",
-            include_source_files=False,
-        ),
+        PexFromTargetsRequest,
+        requirements_pex_from_targets_request(Addresses(tgt.address for tgt in repl.targets)),
+    )
+    sources_request = Get(
+        PythonSourceFiles, PythonSourceFilesRequest(repl.targets, include_files=True)
     )
-    sources_request = Get(PythonSourceFiles, PythonSourceFilesRequest(repl.targets))
-    pex, sources = await MultiGet(pex_request, sources_request)
+    requirements_pex, sources = await MultiGet(requirements_request, sources_request)
     merged_digest = await Get(
-        Digest, MergeDigests((pex.digest, sources.source_files.snapshot.digest))
+        Digest, MergeDigests((requirements_pex.digest, sources.source_files.snapshot.digest))
     )
     return ReplRequest(
         digest=merged_digest,
-        binary_name=pex.output_filename,
+        binary_name=requirements_pex.output_filename,
         env={"PEX_EXTRA_SYS_PATH": ":".join(sources.source_roots)},
     )
 
 
 class IPythonRepl(ReplImplementation):
     name = "ipython"
-    required_fields = (PythonSources,)
 
 
 @rule
 async def create_ipython_repl_request(repl: IPythonRepl, ipython: IPython) -> ReplRequest:
-    pex_request = Get(
+    # Note that unlike above, we get an intermediate PexRequest here (instead of going straight
+    # to a Pex) so that we can capture the interpreter constraints.
```

Comment:
The `unlike above` threw me off a little. Maybe say

> Note that we immediately evaluate this so that we can use the resulting interpreter_constraints for `ipython_request`.